### ISSUE TITLE
Inferred type is ignored on @psalm-type properties

### ISSUE THREAD

**Issue by SCIF** [vimeo/psalm#7780]
https://psalm.dev/r/6a81914751

**Comment by AndrolGenhald** [comment]
This looks like it's more to do with keyed objects, it happens [without `@psalm-type`](https://psalm.dev/r/c7ed0f11ff) as well, and when [switching to an array](https://psalm.dev/r/75f65db16d) it works fine with `@psalm-type`.

**Comment by orklah** [comment]
Yeah, pretty sure we already have a similar issue but I'm not sure how to find it :)

I think the core of the issue is that $a['x'] can be returned by `ExpressionIdentifier::getVarId` while `$a->x` can't

**Comment by SCIF** [comment]
> I think the core of the issue is that $a['x'] can be returned by `ExpressionIdentifier::getVarId` while `$a->x` can't

I don't really understand the logic of ExpressionIdentifier::getVarId() but there are [next lines](https://github.com/vimeo/psalm/blob/master/src/Psalm/Internal/Analyzer/Statements/Expression/ExpressionIdentifier.php#L56-L64).

@orklah , could you advise the direction of further investigation?

**Comment by orklah** [comment]
I looked into it and I was wrong indeed. `getVarId` (or `getExtendedVarId` instead on master) does support $a->x correctly.

Seems like there is a nest of issues here
If we compare those two:
https://psalm.dev/r/d1bb867ae0
https://psalm.dev/r/08bbca1696

Inside the if, it seems like the type of 'some' is correctly propagated for arrays but not for objects. It must be missing some support here. If I had to guess, I'd say it should be somewhere near `Reconciler::reconcileKeyedTypes` but it's not a part of the code I understand fully yet.

I think this would solve the value outside the if too (your initial bug report) because it's basically the same propagation issue.

There is a last bug, `$a->some` become mixed outside of the if, I didn't check that for now
